iib/web/healthcheck.py

```diff
@@ -0,0 +1,138 @@
+        return False, "an error occurred"
+    finally:
+        session.close()
+
+    return True, None
+
+
+def rabbitmq_connection_check(broker_url):
+    """
+    Check if RabbitMQ connection is working.
+
+    :param str broker_url: the AMQP(S) URL to connect to RabbitMQ
+    :return: tuple (ok: bool, reason: str or None)
+    """
+    with kombu.Connection(broker_url) as connection:
+        try:
+            connection.ensure_connection(max_retries=0)
+        except kombu.exceptions.OperationalError as OperError:
+            log.exception("broker connection failed, exception: %s", OperError)
+            return False, "broker connection failed"
+        except kombu.exceptions.KombuError as err:
+            log.error("an error occurred while trying to connect to the RabbitMQ server: %s", err)
+            return False, "an error occurred"
+
+    return True, None
+
+
+def _ping_celery_workers(retries):
+    """
+    Attempt to ping worker, retry on ConnectionError.
+
+    :param int retries: number of retry/retries before returning an empty response
+    :return: dict of {hostname: reply} as returned by ping()
+    """
+    for i in range(retries + 1):
+        if i > 0:
+            time.sleep(0.25 * 2 ** (i - 1))
+        try:
+            replies = app.control.inspect().ping()
+        except ConnectionError:
+            continue
+
+        return replies or {}
+
+    return {}
+
+
+def workers_status(retries=2):
+    """
+    Ping workers, check received replies.
+
+    :param int retries: number of retry/retries on ConnectionError when pinging workers
+    :return: list of workers status information (empty if no replies)
+    """
+    replies = _ping_celery_workers(retries=retries)
+
+    # replies is a dict of {hostname: reply}, convert to a sorted list of [(hostname, reply)]
+    reply_tuples = sorted(replies.items(), key=lambda kv: kv[0])
+    workers = []
+
+    for worker_name, reply in reply_tuples:
+        # the reply is in the following format {'ok': 'pong'}.
+        worker = {"name": worker_name, "available": "ok" in reply}
+
+        if "ok" not in reply:
+            worker["reason"] = reply.get("error", "unknown reason")
+
+        workers.append(worker)
+
+    return workers
+
+
+def status(short=False, worker_ping_retries=2):
+    """
+    Get status of IIB workers and services that IIB depends on.
+
+    :param bool short: raise an error as soon as any problem is found
+    :param int worker_ping_retries: number of retry/retries on ConnectionError when pinging workers
+    :return: dict with the following keys:
+        "services": list of status info for individual services
+        "workers": list of status info for individual workers
+    :raises IIBError: if short is True and a problem is found
+    """
+    services = []
+
+    def add_status(service_name, ok, reason):
+        if short and not ok:
+            raise IIBError(f"{service_name} unavailable: {reason}")
```

Comment:
If a service is not ok, then the other services are never checked. It may be desirable to always provide the status for every service.

Reply:
in case of short is True, it would only show the status in general (Status: health check failed and the reason) - just like the current healthcheck endpoint so nothing would break as it was suggested & mentioned in this PR. 

As this is a short healthcheck, if one of the services is not working, it should show the status as failed. 

If you wanna see the status of all services, you will need to check `healthcheck/details` - it contains a list of the services ( database & rabbitmq ), workers and their status.

Reply:
Fair point. It's best to stop checking the other services in this case. Thanks for the clarification.